hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3147 @@
+    os_semaphore_t disconnectSemaphore_;            /**< Semaphore to wait until connection disconnected. */
+    Vector<BleConnection_t> connections_;           /**< Current on-going connections. */
+    BleObject* bleObj_;                             /**< Pointer to the BLE Object. */
+
+    static const uint32_t CONNECTION_OPERATION_TIMEOUT_MS = 5000;
+
+    BleConnection_t* fetchConnection(uint16_t connHandle);
+    BleConnection_t* fetchConnection(const hal_ble_addr_t* address);
+    void addConnection(const BleConnection_t& connection);
+    void removeConnection(uint16_t connHandle);
+    void initiateConnParamsUpdateIfNeeded(const BleConnection_t* connection);
+    bool isConnParamsFeeded(const hal_ble_conn_params_t* params) const;
+    static void toPlatformConnParams(ble_gap_conn_params_t* params, const hal_ble_conn_params_t* halConnParams);
+    static void toHalConnParams(hal_ble_conn_params_t* halConnParams, const ble_gap_conn_params_t* params);
+    static void onConnParamsUpdateTimerExpired(os_timer_t timer);
+    static void processConnectionEvents(const ble_evt_t* event, void* context);
 };
 
-// Current profile
-Profile g_profile = {};
+class BleObject::GattServer : public GattBase {
+public:
+    GattServer(BleObject* obj);
+    ~GattServer() = default;
+    int init(void);
+    int addService(uint8_t type, const hal_ble_uuid_t* uuid, uint16_t* handle);
+    int addCharacteristic(uint16_t svcHandle, uint8_t properties, const hal_ble_uuid_t* uuid, const char* desc, hal_ble_char_handles_t* handles);
+    int addDescriptor(uint16_t charHandle, const hal_ble_uuid_t* uuid, uint8_t* descriptor, uint16_t len, uint16_t* handle);
+    size_t setValue(uint16_t handle, const uint8_t* buf, size_t len);
+    size_t getValue(uint16_t handle, uint8_t* buf, size_t len);
+
+private:
+    class BleCharacteristic {
+        public:
+            uint8_t properties;
+            uint16_t svcHandle;
+            hal_ble_char_handles_t handles;
+            Vector<uint16_t> cccdConnections;
+    };
+
+    static const uint32_t BLE_HVX_PROCEDURE_TIMEOUT_MS = 5000;
+
+    os_semaphore_t hvxSemaphore_;                   /**< Semaphore to wait until the HVX operation completed. */
+    Vector<uint16_t> services_;                     /**< Added services. */
+    Vector<BleCharacteristic> characteristics_;     /**< Added characteristic. */
+    BleObject* bleObj_;                             /**< Pointer to the BLE Object. */
+    AtomicAllocedPool charValuePool_;               /**< Pool to allocate memory for characteristic value. */
+    AtomicAllocedPool gattsDataRecPool_;            /**< Pool to allocate memory for received data. */
+
+    static const uint16_t GATT_SERVER_DATA_REC_POOL_SIZE = 1024;
 
-// Connection state
-volatile Conn g_conn = {
-    .handle = BLE_CONN_HANDLE_INVALID,
-    .paramUpdateAttempts = 0
+    bool findService(uint16_t handle) const;
+    BleCharacteristic* findCharacteristic(uint16_t handle);
```

Comment:
It might be better to have separate methods like `findCharacteristicByAParticularHandle()`

Reply:
Attribute handles belonging to a characteristic are continuous, so by passing any kind of attribute handle (no matter CCCD handle, value handle or user description handle, etc.) can find parent characteristic. Otherwise, there will be many methods. I think that is unnecessary.

Reply:
Ok, if that is guaranteed by the API provided by Nordic. Let's at least leave a comment in that case.

Reply:
This is not related to the Nordic API. Each attribute handle is unique. The implementation of this method will check whether the given attribute handle matches any of the attribute handles within a characteristic, once found then return it.